src/type/s2k/argon2.js

```diff
@@ -0,0 +1,137 @@
+import defaultConfig from '../../config';
+import enums from '../../enums';
+import util from '../../util';
+import crypto from '../../crypto';
+
+const ARGON2_TYPE = 0x02; // id
+const ARGON2_VERSION = 0x13;
+const ARGON2_SALT_SIZE = 16;
+
+export class Argon2OutOfMemoryError extends Error {
+  constructor(...params) {
+    super(...params);
+
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, Argon2OutOfMemoryError);
+    }
+
+    this.name = 'Argon2OutOfMemoryError';
+  }
+}
+
+// cache argon wasm module
+let loadArgonWasmModule;
+let argon2Promise;
+// reload wasm module above this treshold, to deallocated used memory
+const ARGON2_WASM_MEMORY_THRESHOLD_RELOAD = 2 << 19;
+
+class Argon2S2K {
+  /**
+  * @param {Object} [config] - Full configuration, defaults to openpgp.config
+  */
+  constructor(config = defaultConfig) {
+    const { passes, parallelism, memoryExponent } = config.s2kArgon2Params;
+
+    this.type = 'argon2';
+    /**  @type {Uint8Array} 16 bytes of salt */
+    this.salt = null;
+    /** @type {Integer} number of passes */
+    this.t = passes;
+    /** @type {Integer} degree of parallelism (lanes) */
+    this.p = parallelism;
+    /** @type {Integer} exponent indicating memory size */
+    this.encodedM = memoryExponent;
+  }
+
+  generateSalt() {
+    this.salt = crypto.random.getRandomBytes(ARGON2_SALT_SIZE);
+  }
```

Comment:
I know the existing S2K class did something similar, but rather than setting the `t`, `p` and `encodedM` parameters in the constructor, and then optionally overwriting them in `read()`, I would prefer to make it explicit when we actually use those. You could have a `generateParams(config)` function instead of a `generateSalt()`, and set both the salt and the other parameters there.
(And similarly in the `GenericS2K` class.)

Reply:
I am not sure about this, with packets we also set default values in constructor, even tho they can be overwritten when reading.
API-wise, since the s2k instance is not really meant to be reused, having a function that re-generates potentially different configuration could be confusing. The salt is fundamentally different from the rest of the params, since it's meant to change for each S2K instance. Also, especially if we add `newS2KFromConfig`, selectively applying the config seems unexpected.